source/extensions/http/header_validators/envoy_default/path_normalizer.cc

```diff
@@ -0,0 +1,400 @@
+    if (iter == end) {
+      return {PercentDecodeResult::Invalid};
+    }
+
+    char nibble = *iter;
+    if (!isxdigit(*iter)) {
+      return {PercentDecodeResult::Invalid};
+    }
+
+    // normalize
+    nibble = nibble >= 'a' ? nibble ^ 0x20 : nibble;
+    *iter = nibble;
+
+    // decode
+    int factor = i == 0 ? 16 : 1;
+    ch += factor * (nibble >= 'A' ? (nibble - 'A' + 10) : (nibble - '0'));
+  }
+
+  if (testChar(kUnreservedCharTable, ch)) {
+    // Based on RFC, only decode characters in the UNRESERVED set.
+    return {PercentDecodeResult::Decoded, ch};
+  }
+
+  if (ch == '/') {
+    // We decoded a slash character and how we handle it depends on the active configuration.
+    switch (config_.uri_path_normalization_options().path_with_escaped_slashes_action()) {
+    case HeaderValidatorConfig_UriPathNormalizationOptions::IMPLEMENTATION_SPECIFIC_DEFAULT:
+      ABSL_FALLTHROUGH_INTENDED;
+    case HeaderValidatorConfig_UriPathNormalizationOptions::KEEP_UNCHANGED:
+      // default implementation: normalize the encoded octet and accept the path
+      return {PercentDecodeResult::Normalized};
+
+    case HeaderValidatorConfig_UriPathNormalizationOptions::REJECT_REQUEST:
+      // Reject the entire request
+      return {PercentDecodeResult::Reject};
+
+    case HeaderValidatorConfig_UriPathNormalizationOptions::UNESCAPE_AND_FORWARD:
+      // Decode the slash and accept the path.
+      return {PercentDecodeResult::Decoded, ch};
+
+    case HeaderValidatorConfig_UriPathNormalizationOptions::UNESCAPE_AND_REDIRECT:
+      // Decode the slash and response with a redirect to the normalized path.
+      return {PercentDecodeResult::DecodedRedirect, ch};
+
+    default:
+      // This should never occur but it's here to make the compiler happy because of the extra
+      // values added by protobuf.
+      break;
+    }
+  }
+
+  // The octet is a valid encoding but it wasn't be decoded because it was outside the UNRESERVED
+  // character set.
+  return {PercentDecodeResult::Normalized};
+}
+
+/*
+ * Find the start index of the previous segment within the path. The previous segment starts at the
+ * first non-slash character after the preceding slash. For example:
+ *
+ *   path = "/hello/world/..";
+ *                  ^    ^-- iterator
+ *                  |-- start of previous segment
+ *
+ * The ``begin`` iterator is returned on error.
+ */
+std::string::iterator findStartOfPreviousSegment(std::string::iterator iter,
+                                                 std::string::iterator begin) {
+  bool seen_segment_char = false;
+  for (; iter != begin; --iter) {
+    if (*iter == '/' && seen_segment_char) {
+      ++iter;
+      return iter;
+    }
+
+    if (*iter != '/' && !seen_segment_char) {
+      seen_segment_char = true;
+    }
+  }
+
+  if (seen_segment_char) {
+    ++begin;
+  }
+
+  return begin;
+}
+
+PathNormalizer::PathNormalizationResult
+PathNormalizer::normalizePathUri(RequestHeaderMap& header_map) const {
```

Comment:
This introduces breaking change, although it is unclear to me if the impact is meaningful. In this implementation, slash merging and decoding of %2F is only done if path_normalization is enabled. In existing implementation these options are independent of each other. I.e. it is possible to only decode %2F or just merge slashes, without making other transformations. 

I think it is unlikely that someone would enable merging slashes without enabling path normalization. 

Please add a comment that the existing behavior has changed and we can evaluate if we need to do anything about it later.

Reply:
Done. I added a TODO referencing a new issue: #23887